archetype engine v2 design doc (#340)
oracle/helidon-build-tools: docs-internal/archetype-v2.md

```xml
<step label="Application Type">
    <!-- ... -->
</step>
```

### Optional Steps

Optional steps must have inputs with defaults. An optional step with non default inputs is invalid and should be
 validated as an error.

Customization of features can be modeled using an optional step declared as the last child of the step to be
 customized. The order is the declaration order, the enclosing steps must declare the optional step used for
 customization carefully.

E.g.
- `se/se.xml` executes `se/hello-world/hello-world-se.xml` AND THEN executes `common/customize-project.xml`
- `se/hello-world/hello-world-se.xml` executes `se/hello-world/customize-hello-world.xml` as the very last step.

The declaration order for the two customization steps is first `customize-hello-world` and then `customize-project`.

When interpreting the flow, the next steps are always computed. If the next steps are all optional, it is possible to
 skip them and generate the project.

An optional step requires no action other than continuing to the next step (unless this is the very last step).

## Flow Context

> The flow context is a tree of resolved inputs that represents the choices. The tree nodes represent choices that
>  are a pair of an input path its resolved value.

Reviewer: ```suggestion
The flow context is a tree of _resolved_ inputs, where each node has a name (and so can be addressed via a path) and contains the resolved value and any children.
```

> The context may be populated with various mechanisms:

Reviewer: ```suggestion
The context may be populated _without_ user input to constrain the choices presented to the user, causing the interpreter to skip over the associated inputs as if the user had already made the choice. Several mechanisms can be used for this purpose:
```

> - `<context>` directive
> - query parameters
> - CLI option

Reviewer: ```suggestion
- the `<context>` directive
- URI query parameters passed when starting the script interpreter
- CLI options passed when starting the script interpreter
```

> 

Reviewer: Insert something here about fully specifying the presets such that no choices remain.

> E.g.

Reviewer: Need to describe this example tree (e.g. why are some values read-only).

If all nodes share a common parent hierarchy, a common prefix can be configured to indicate the prefix of absolute
 paths that may be omitted.

E.g. Consider the following absolute paths with a common prefix `flavor.base`:
```
flavor // resolved as flavor
favor // resolved as flavor.base.favor
flavor.base // resolved as flavor.base
flavor.foo // INVALID
flavor.base.media-type.provider // resolved flavor.base.media-type.provider
security.authentication.provider // resolved flavor.base.security.authentication.provider
```

When used within the descriptors, paths are always relative to their current context, unless they start with
 `ROOT.` or `PARENT.`:
```
${ROOT.flavor.base.security.authentication.provider} // absolute, resolved as flavor.base.security.authentication.provider
${ROOT.security.authentication.provider} // absolute, resolved as flavor.base.security.authentication.provider
${provider} // resolved as flavor.base.security.provider if current is flavor.base
${PARENT.security.provider} // resolved as flavor.base.security.provider if the parent of current is flavor.base
```

Invalid paths must be validated at build time and reported as errors.

### Flow Context Directive

> The `<context>` directive can be used to pre-set the value of a given input in the flow context. The values set

Reviewer: ```suggestion
The `<context>` directive can be used to preset the value of a given input in the flow context. The values set
```

- `${security}`

Inline search and replace regular expressions are also supported:
- `${package/\./\/}`

This can be used to add data in the template data model, or to set default values for other inputs.

### Flow Context Path Expressions

Flow context path expressions are boolean expressions that can be used to query the flow context.

Values can be either flow context path values or literal, and are only of the following types: 
 - Boolean: `true` or `false`
 - Text: `'foo'`
 - Array: `['foo', 'bar']`

The following operators are supported:
 - `&&`: logical AND
 - `||`: logical OR
 - `!`: logical negation
 - `contains`: array contains
 - `==`: equality

```
${security}
${media-support.json.provider} == 'jackson'
${security} && ${media-support}
${security} || ${media-support}
${security.authentication.provider} contains 'basic-auth'
!(${security} && ${media-support}) || ${health}
```

Flow context path expressions are supported in the following elements:
 - anything under `<output if="expr">`
 - `<step if="expr">`

### External Flow Context Values

Values in the flow context can also be set externally using CLI options or URI query parameters. The flow context

>  nodes are identified using absolute paths, thus the common prefix can be omitted (`PARENT.` and `ROOT.` are not
>  allowed).

Reviewer: ```suggestion
 nodes must be identified using absolute paths; the common prefix can be omitted, but `PARENT.` and `ROOT.` are not allowed.
```

When declaring external flow context values, gating input option values can be inferred.
 E.g. `media-support.json.provider=jackson` implies both `media-support=true` and `media-support.json=true`.

Query parameters:
```
?media-support.json.provider=jackson&security.authentication.provider=basic-auth,digest-auth&health=false&project-name=my-super-project
```

Properties:
```properties
media-support.json.provider=jackson
security.authentication.provider=basic-auth,digest
health=false
project-name=my-super-project
```

The `.helidon` file generated with the project will include the user's choices. A prefix need to be used to identify the
 choices properties, e.g. `FLOW.media-support.json.provider`.

CLI options:
```
helidon init \
    --input media-support.provider=jackson \
    --input security.authentication.provider=basic-auth,digest
    --input health=false
    --input project-name=my-super-project
```

> ## Help text

Reviewer: ```suggestion
## Help Text
```
This same case issue appears a number of times below.

            </plugin>
        </plugins>
    </build>
    <dependencies>
        <dependencies>
            <!-- main archetype -->
            <dependency>
                <groupId>io.helidon.archetypes</groupId>
                <artifactId>helidon-archetype</artifactId>
            </dependency>
            <!-- archetype engine -->
            <dependency>
                <groupId>io.helidon.build-tools.archetype</groupId>
                <artifactId>helidon-archetype-engine</artifactId>
            </dependency>
            <!-- mvn:// URL support -->
            <dependency>
                <groupId>io.helidon.build-tools.archetype</groupId>
                <artifactId>helidon-archetype-maven-url</artifactId>
            </dependency>
        </dependencies>
    </dependencies>
</project>
``` 

## Mock-ups

### UI wizard

- Select:

```
(1) Application Type
 |
 | Select a type of application:
 |  (   ) Bare
 |  ( X ) Hello World
 |  (   ) Database
 |
 | [CONTINUE]
-----------------------------------------
```

- Option gating a step:

```
(2) Kubernetes
 |
 | [  ] Kubernetes support
 |
 | [CONTINUE]
-----------------------------------------
```

> Such a step requires not user action. Clicking the check box expands nested inputs:

Reviewer: ```suggestion
Such a step does not require user action (so the CONTINUE button is active), but allows it; clicking the check box expands nested inputs:
```

```
(2) Kubernetes
 |
 | [ x ] Kubernetes support
 |   |- [ x ] add a service

>  |   |- [ x ] add an Istio sidecar

Reviewer: I think it would be clearer if these were unchecked.
